pyterrier/pipelines.py

```diff
@@ -147,33 +161,27 @@ def _run_and_evaluate(
             metrics,
             rev_mapping,
             num_q,
-            perquery)
+            perquery,
+            backfill_qids)
     else:
         #transformer, evaluate queries in batches
         assert batch_size > 0
         starttime = timer()
         results=[]
-        evalMeasuresDict={}
         for i, res in enumerate( system.transform_gen(topics, batch_size=batch_size)):
             if len(res) == 0:
                 raise ValueError("batch of %d topics, but no results received in batch %d from %s" % (batch_size, i, str(system) ) )
             endtime = timer()
             runtime += (endtime - starttime) * 1000.
-            localEvalDict = _ir_measures_to_dict(
-                ir_measures.iter_calc(metrics, qrels, res.rename(columns=_irmeasures_columns)),
-                metrics,
-                rev_mapping,
-                num_q,
-                True)
-            evalMeasuresDict.update(localEvalDict)
+            results.append(res)
             starttime = timer()
-        if not perquery:
-            aggregators = {rev_mapping.get(m, str(m)): m.aggregator() for m in metrics}
-            for q in evalMeasuresDict:
-                for metric in metrics:
-                    s_metric = rev_mapping.get(metric, str(metric))
-                    aggregators[s_metric].add(evalMeasuresDict[q][s_metric])
-            evalMeasuresDict = {m: agg.result() for m, agg in aggregators.items()}
+        evalMeasuresDict = _ir_measures_to_dict(
```

Comment:
This change may be controversial. When `batch_size` is set, it no longer evaluates after each batch, but rather keeps all the results and evaluates everything at the end.

Alternatively, I could update so it filters the qrels each time.